### PR TITLE
Adding support for `.ejs` files

### PR DESCRIPTION
**Pull request by Robbie-Cook**
Also fixes [#15 ],[#14 ].

`@bookshop/ejs-engine` needs to be directly linked to, but it needs to exist first. I plan to do this after merge.

~~Sorry about the formatting of the document -- I accidentally formatted it halfway through.~~
~~P.S. The `.ejs` engine is transpiled because ejs needs to be browserified -- it has `require(fs)`'s in it.~~

### CODE REVIEW

File: examples/example-bookshop/package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "@babel/core": "^7.11.6",
     "@babel/plugin-transform-runtime": "^7.11.5",
     "@babel/preset-env": "^7.11.5",
+    "@bookshop/ejs-engine": "file:../../engines/ejs-engine",
```
Comment (Robbie-Cook) [inline examples/example-bookshop/package.json:24]: This needs to be linked directly to the package when the package exists

File: loader/package.json [hunk 2]
```diff
@@ -17,6 +17,7 @@
     "@bookshop/jekyll-engine": "^1.5.3",
     "@bookshop/svelte-engine": "^1.5.3",
     "@bookshop/underscore-engine": "^1.5.3",
+    "@bookshop/ejs-engine": "file:../engines/ejs-engine",
```
Comment (Robbie-Cook) [inline loader/package.json:20]: This also needs to be directly linked to the package when it exists

**Comment by Robbie-Cook** [comment]
@bglw

**Review by bglw — CHANGES_REQUESTED** [review]
It would be great if we can avoid having to transpile this engine.

If you include the client build (https://npmcdn.com/ejs@3.1.5/ejs.js) then we can run it in the browser without any transpilation. I don't think there's an npm module with the client build, but that file looks requirable if it were dropped in. 

You'll also need to add support for includes inside ejs files. You can see how to handle includes on their Github: https://github.com/mde/ejs#caveats

So on catching an include you'll need to do the network request to get a file and return it (similar to underscore-engine behaviour)

**Comment by Robbie-Cook** [comment]
Great! Thank you for your review!

> It would be great if we can avoid having to transpile this engine.
> 
> If you include the client build ([npmcdn.com/ejs@3.1.5/ejs.js](https://npmcdn.com/ejs@3.1.5/ejs.js)) then we can run it in the browser without any transpilation. I don't think there's an npm module with the client build, but that file looks requirable if it were dropped in.

I'll do this.

However, I think this is potentially problematic. Transpiling allows us to keep up to date with the latest `ejs` package, but maunally copying a script so `require` will work natively makes it harder to update. 

There are also other advantages of transpiling (e.g. ES2020 support for older node LTS versions) which I am sure you are aware of.

> 
> You'll also need to add support for includes inside ejs files. You can see how to handle includes on their Github: [mde/ejs#caveats](https://github.com/mde/ejs#caveats)

> 
> So on catching an include you'll need to do the network request to get a file and return it (similar to underscore-engine behaviour)

Great! Thanks for this. I will start adding this now. Good catch!

**Comment by bglw** [comment]
> Great! Thank you for your review @bglw !
> 
> > It would be great if we can avoid having to transpile this engine.
> > If you include the client build ([npmcdn.com/ejs@3.1.5/ejs.js](https://npmcdn.com/ejs@3.1.5/ejs.js)) then we can run it in the browser without any transpilation. I don't think there's an npm module with the client build, but that file looks requirable if it were dropped in.
> 
> I'll do this.
> 
> However, I think this is potentially problematic. Transpiling allows us to keep up to date with the latest `ejs` package, but maunally copying a script so `require` will work natively makes it harder to update.
> 
> There are also other advantages of transpiling (e.g. ES2020 support for older node LTS versions) which I am sure you are aware of.

Let's do this more problematic way for now, just so we don't have to deal with transpilation for the bookshop refactor that's in the pipes. After that, we can revisit this setup.
> 
> > You'll also need to add support for includes inside ejs files. You can see how to handle includes on their Github: [mde/ejs#caveats](https://github.com/mde/ejs#caveats)
> 
> > So on catching an include you'll need to do the network request to get a file and return it (similar to underscore-engine behaviour)
> 
> Great! Thanks for this. I will start adding this now. Good catch!
🤙

**Comment by Robbie-Cook** [comment]
Hey @bglw , have removed webpack.

Just working on the `include` behaviour now :)

**Comment by Robbie-Cook** [comment]
Hey @bglw , includes now work in storybook! 

The syntax matches the `ejs` spec e.g. `include('components/button/button.ejs', { props: { text: props.text, type } });`, 

This should be good to go, I think!

**Comment by Robbie-Cook** [comment]
Also, cleaned up the diff a bit by undoing my prettier format, so should be nicer to read

File: loader/src/index.js [hunk 3]
```diff
@@ -119,22 +121,18 @@ module.exports = function(source) {
 
   let {component, title, full} = getTitleFromPath(this);
 
-  if (hasMultipleFrameworks) {
-    data.defaults = {
-      framework: frameworks[0] || 'jekyll',
-      ...data.defaults
-    }
-  }
+  data.defaults = {
```
Comment (Robbie-Cook) [inline loader/src/index.js:124]: This fixes #14

File: engines/ejs-engine/src/ejs.js [hunk 4]
```diff
@@ -0,0 +1,1661 @@
+(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.ejs = f()}})(function(){var define,module,exports;return (function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
```
Comment (Robbie-Cook) [inline engines/ejs-engine/src/ejs.js:1]: You can ignore this file -- it is the ejs browsified renderer

**Review by bglw — COMMENTED** [review]
- Just checking - you access variables through a key `props`? So you would expect `<%= include.text %>` rather than `<%= text %>`
- Curious about the decision to override the include variable rather than using the [include callback](https://github.com/mde/ejs#caveats) feature ejs has?
- If you do intend to use the override instead, looks like you will also need to provide it in an included file. Looking at it I don't think including a file that includes a file will work as is.

**Comment by Robbie-Cook** [comment]
Thanks @bglw 

> * Just checking - you access variables through a key `props`? So you would expect `<%= include.text %>` rather than `<%= text %>`

I think that you don't necessarily need to key the passed data to `props` - this is just mapped from the storybook loader render function (operates the same as your `underscore` engine). Here is the syntax for an include in a `.ejs` file: 

image

> * Curious about the decision to override the include variable rather than using the [include callback](https://github.com/mde/ejs#caveats) feature ejs has?

> * If you do intend to use the override instead, looks like you will also need to provide it in an included file. Looking at it I don't think including a file that includes a file will work as is.

As is, including a file that includes a file seems to be working fine.
Here is an example of a file that includes a button component, and the button component includes an input component i.e.:

overview
	-> button.ejs
		-> filter.ejs

: 

image

Is this okay? The current solution was just the easiest one, and it seems to work fine for the storybook. 
Ofc, the loader for the app will have to be different

**Review by bglw — APPROVED** [review]
Cool, looks good then

**Comment by Robbie-Cook** [comment]
@bglw We've decided to make the `props` the only data passed in, so the new syntax for importing will be:

```ejs
<%- include('components/button/button.ejs', { text: props.text, type }); %>
```.

It is not as general as it could be, the benefits are component consistency and succintness

**Pull request merged**